### PR TITLE
refactor(db/schema): improve expression validation

### PR DESCRIPTION
**Pull request by chronolaw**
### Summary

Why:
- The atc-router library updates its build files, we should update too
- The new `validate` API in atc-router could improve the validation performance.
- The new rust.regex crate supports special escape, which will case some tests cases fail
- The code in the schema is a bit hard to understand

When the atc-router releases a new version, this PR will be ready to merge.

KAG-1524

### Checklist

- [x] The Pull Request has tests
- [x] There's an entry in the CHANGELOG
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Full changelog

* localize atc-router module
* cache atc-router matcher
* rename to `validate_entity_by_expression` and change entrance
* constant `PATH_V1_DEPRECATION_MSG `
* check paths only in `traditional_compatible`
* fix some regex path test cases for rust regex bump to 1.8
* fix bazel build for latest atc-router updating

### CODE REVIEW

File: kong/db/schema/entities/routes.lua [hunk 1]
```diff
@@ -73,6 +90,55 @@ if kong_router_flavor == "expressions" then
...
+
+  if kong_router_flavor == "traditional" then
+    DEPRECATION_MSG = "path_handling='v1' is deprecated and will be removed in future version, " ..
+                      "please use path_handling='v0' instead"
+
+  elseif kong_router_flavor == "traditional_compatible" then
+    DEPRECATION_MSG = "path_handling='v1' is deprecated and will not work under traditional_compatible " ..
+                      "router_flavor, please use path_handling='v0' instead"
+  end
+
```
Comment (StarlightIbuki) [inline kong/db/schema/entities/routes.lua:103]: This does not seem as straightforward as the original code: is it already deprecated if we are using traditional?

Please consider renaming the `DEPRECATION_MSG` to a more specific name, like `PATH_V1_DEPRECATION_MSG`.

**Review by StarlightIbuki — COMMENTED** [review]
Good to go after renaming is done.

File: kong/db/schema/entities/routes.lua [hunk 2]
```diff
@@ -2,18 +2,35 @@ local typedefs = require("kong.db.schema.typedefs")
...
+  local type = type
 
-local function validate_expression(id, exp)
   local r = router.new(CACHED_SCHEMA)
 
-  local res, err = r:add_matcher(0, id, exp)
-  if not res then
-    return nil, "Router Expression failed validation: " .. err
+  validate_expression = function(entity)
```
[the remark this answers is not in the document]
Comment (chronolaw) [reply kong/db/schema/entities/routes.lua:16]: rename to `validate_entity_by_expression`

File: kong/db/schema/entities/routes.lua (hunk 1 above)
Comment (chronolaw) [reply kong/db/schema/entities/routes.lua]: done

File: kong/db/schema/entities/routes.lua [hunk 3]
```diff
@@ -2,18 +2,35 @@ local typedefs = require("kong.db.schema.typedefs")
...
-  local res, err = r:add_matcher(0, id, exp)
-  if not res then
-    return nil, "Router Expression failed validation: " .. err
+  local r = router.new(CACHED_SCHEMA, 1)
+
+  validate_entity_by_expression = function(entity)
+    local id = entity.id
+    local exp = entity.expression or get_expression(entity)
+
+    local res, err = r:add_matcher(0, id, exp)
```
Comment (chronolaw) [inline kong/db/schema/entities/routes.lua:20]: Could we add a new API like `r:validate()` in atc_router? 

In my opinion, `r:add_matcher` is not the best idea to validate expression.

Comment (chronolaw) [reply kong/db/schema/entities/routes.lua:20]: We should implement this API in atc-router's Lua code.

Comment (chronolaw) [reply kong/db/schema/entities/routes.lua:20]: need to refactor after https://github.com/Kong/atc-router/pull/39 merged.

**Comment by dndx** [comment]
@chronolaw Is this ready for review?

**Comment by dndx** [comment]
Blocked by https://github.com/Kong/atc-router/pull/39.

**Comment by mikefero** [comment]
@hbagdi Can you elaborate on the version compatibility concerns here as this refactor looks to keep the schema intact?

**Comment by hbagdi** [comment]
No concerns @mikefero.

**Comment by mikefero** [comment]
> No concerns @mikefero.

None; the schema is intact (just moved) and only impacts validation of the expression router. All good here.

**Comment by hbagdi** [comment]
@chronolaw Any blockers on this PR?

**Comment by mikefero** [comment]
> > No concerns @mikefero.
> 
> None; the schema is intact (just moved) and only impacts validation of the expression router. All good here.

@hbagdi Taking this statement back now, it looks like we are removing a field.

@chronolaw What is the reasoning for removing a field in [f771d3f](https://github.com/Kong/kong/pull/10100/commits/f771d3f2c0207b909f8bf2dd8f1c2d97d0d9d734)? This will have a negative impact on version compatibility with older data planes.

**Comment by chronolaw** [comment]
> > > No concerns @mikefero.
> > 
> > 
> > None; the schema is intact (just moved) and only impacts validation of the expression router. All good here.
> 
> @hbagdi Taking this statement back now, it looks like we are removing a field.
> 
> @chronolaw What is the reasoning for removing a field in [f771d3f](https://github.com/Kong/kong/pull/10100/commits/f771d3f2c0207b909f8bf2dd8f1c2d97d0d9d734)? This will have a negative impact on version compatibility with older data planes.

No, the field `id` is only be passed to validation, and we don't use it any more with new api of atc-router.

File: kong/db/schema/entities/routes.lua [hunk 4]
```diff
@@ -2,18 +2,30 @@ local typedefs = require("kong.db.schema.typedefs")
...
+  local get_expression = require("kong.router.compat").get_expression
 
-local function validate_expression(id, exp)
-  local r = router.new(CACHED_SCHEMA)
+  local type = type
 
-  local res, err = r:add_matcher(0, id, exp)
-  if not res then
-    return nil, "Router Expression failed validation: " .. err
+  validate_entity_by_expression = function(entity)
```
Comment (dndx) [inline kong/db/schema/entities/routes.lua:14]:
```suggestion
  validate_traditional_compatiable_route = function(entity)
```

Might be better illustration of the purpose of this function.

Comment (chronolaw) [reply kong/db/schema/entities/routes.lua]: rename to `validate_route`

**Pull request merged**